File: frontend/lib/models/transaction.dart
class Transaction {
  final int id;
  final int accountId;
  final double amount;
  final DateTime effectiveDate;
  final String displayName; // Backend's computed display_name
  final String? memo;
  final int? subcategoryId;
  final bool pending;
  final DateTime createdAt;
  final bool isTransfer;
  final int? transferAccountId;

  // ML prediction fields
  final int? predictedSubcategoryId;
  final double? predictedConfidence;

  Transaction({
    required this.id,
    required this.accountId,
    required this.amount,
    required this.effectiveDate,
    required this.displayName,
    this.memo,
    this.subcategoryId,
    required this.pending,
    required this.createdAt,
    this.isTransfer = false,
    this.transferAccountId,
    this.predictedSubcategoryId,
    this.predictedConfidence,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'],
      accountId: json['account_id'],
      amount: json['amount'].toDouble(),
      effectiveDate: DateTime.parse(json['effective_date']),
      displayName: json['display_name'],
      memo: json['memo'],
      subcategoryId: json['subcategory_id'],
      pending: json['pending'],
      createdAt: DateTime.parse(json['created_at']),
      isTransfer: json['is_transfer'] ?? false,
      transferAccountId: json['transfer_account_id'],
      predictedSubcategoryId: json['predicted_subcategory_id'],
      predictedConfidence: json['predicted_confidence']?.toDouble(),
    );
  }

  /// Returns true if transaction has ML prediction but needs manual review
  /// (confidence < 80%)
  bool get needsReview {
    return subcategoryId == null &&
        predictedSubcategoryId != null &&
        predictedConfidence != null &&
        predictedConfidence! < 0.8;
  }

  /// Returns true if transaction was auto-assigned by ML (confidence >= 80%)
  bool get wasAutoAssigned {
    return subcategoryId != null &&
        predictedSubcategoryId == subcategoryId &&
        predictedConfidence != null &&
        predictedConfidence! >= 0.8;
  }
}

File: server/app/models/transaction.py
from sqlalchemy import Column, Integer, String, Float, DateTime, ForeignKey, Boolean, JSON
from sqlalchemy.orm import relationship
from app.core.database import Base
from datetime import datetime
from typing import Optional

class Transaction(Base):
    __tablename__ = "transactions"
    id = Column(Integer, index=True, primary_key=True)
    
    account_id = Column(String, ForeignKey("accounts.id"))
    transaction_id = Column(String)
    
    # Basic transaction info
    amount = Column(Float)
    posted = Column(DateTime)
    transacted_at = Column(DateTime, nullable=True)
    name = Column(String)  # Raw transaction name
        
    # Transaction metadata
    pending = Column(Boolean, default=False)

    # Transfer detection - important for credit card payments, account transfers
    is_transfer = Column(Boolean, default=False)
    transfer_account_id = Column(Integer, ForeignKey("accounts.id"), nullable=True)  # Link to other account
    transfer_transaction_id = Column(Integer, nullable=True)  # Plaid's transfer matching ID
    
    # Location (store as JSON)
    location = Column(JSON, nullable=True)
    
    # ML predicted category (for your custom categorization)
    predicted_category = Column(String, nullable=True)  # Legacy field - not used
    predicted_confidence = Column(Float, nullable=True)  # Confidence score 0.0-1.0
    predicted_subcategory_id = Column(Integer, ForeignKey("subcategories.id"), nullable=True)  # ML prediction
    
    # User-defined budget category (subcategory contains the parent category)
    subcategory_id = Column(Integer, ForeignKey("subcategories.id"), nullable=True)
    
    # User-defined memo/note
    memo = Column(String, nullable=True)
    
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationship
    account = relationship("Account", back_populates="transactions", foreign_keys=[account_id])
    transfer_account = relationship("Account", foreign_keys=[transfer_account_id])
    merchants = relationship("Merchant", secondary="transaction_merchants", back_populates="transactions")
    subcategory = relationship("Subcategory", back_populates="transactions", foreign_keys=[subcategory_id], overlaps="predicted_subcategory")
    predicted_subcategory = relationship("Subcategory", foreign_keys=[predicted_subcategory_id], overlaps="subcategory,transactions")
    
    @property
    def merchant_name(self) -> Optional[str]:
        """Get merchant name(s) filtered by confidence level. Concatenates multiple merchants."""
        if self.merchants:
            # Filter merchants by confidence level and collect names
            merchant_names = [
                merchant.name
                for merchant in self.merchants
                if merchant.confidence_level and merchant.confidence_level.upper() != 'LOW'
            ]
            if merchant_names:
                return ', '.join(merchant_names)
        return None